### PR TITLE
Shielded hardsuit armor

### PR DESCRIPTION
**Pull request by imtakingabreakdontatme**
Shielded hardsuit armour will block all hits until it runs out of charges.

After a time of not being shot it will begin to recharge (being shot again interrupts it).

Number of charges, how quickly these charges regenerate, and how long of a delay until they regenerate are all tweakable via the objects vars.

@RemieRichards I think you said you did overlays so if I'd appreciate it if you could please help me properly update a shield overlay on human mobs depending on whether it has charge or not.

### CODE REVIEW

**Comment by Zilenan91** [comment]
Is this obtainable normally or just admin spawn currently

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 1]
```diff
@@ -430,3 +430,74 @@
...
+
+/////////////SHIELDED//////////////////////////////////
+
+/obj/item/clothing/suit/space/hardsuit/shielded
+	name = "shielded hardsuit"
+	desc = "A hardsuit with built in energy shielding. Will rapidly recharge when not under fire."
+	icon_state = "hardsuit-hos"
+	helmettype = /obj/item/clothing/head/helmet/space/hardsuit/security/hos
+	var/current_charges = 3
+	var/max_charges = 3 //How many charges total the shielding has
```
Comment (Iamgoofball) [inline code/modules/clothing/spacesuits/hardsuit.dm]: i feel like this needs to be like 5 to be good enough

Comment (KorPhaeron) [inline code/modules/clothing/spacesuits/hardsuit.dm]: I wasn't really planning on using the parent type anyway. Also blocking 5 tasers is a ton.

@Zilenan91 

Admin only

**Comment by RemieRichards** [comment]
```
/obj/item/clothing/suit/space/hardsuit/shielded/worn_overlays(isinhands)
    . = list()
    if(!isinhands)
        . += image(icon = 'SHIELD_ICON.dmi',icon_state = "SHIELD_STATE")
```

that's it

**Comment by KorPhaeron** [comment]
Thank you very much

File: code/modules/clothing/spacesuits/hardsuit.dm (hunk 1 above)
Comment (Dorsisdwarf) [inline code/modules/clothing/spacesuits/hardsuit.dm]: Yeah, 3 hits of 100% all projectile block is plenty.

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 2]
```diff
@@ -430,3 +430,76 @@
...
+	var/recharge_rate = 1 //How quickly the shield recharges once it starts charging
+
+
+/obj/item/clothing/suit/space/hardsuit/shielded/hit_reaction(mob/living/carbon/human/owner, attack_text)
+	if(current_charges > 0)
+		var/datum/effect_system/spark_spread/s = new
+		s.set_up(2, 1, src)
+		s.start()
+		owner.visible_message("The [attack_text] sparks harmlessly off [owner]'s shields!")
+		current_charges --
```
Comment (RemieRichards) [inline code/modules/clothing/spacesuits/hardsuit.dm]: typically there's no space between the var and the decrement operator.

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 3]
```diff
@@ -430,3 +430,76 @@
...
 	armor = list(melee = 45, bullet = 25, laser = 30, energy = 10, bomb = 25, bio = 100, rad = 50)
 	helmettype = /obj/item/clothing/head/helmet/space/hardsuit/security/hos
+
+
+
+/////////////SHIELDED//////////////////////////////////
+
+/obj/item/clothing/suit/space/hardsuit/shielded
+	name = "shielded hardsuit"
+	desc = "A hardsuit with built in energy shielding. Will rapidly recharge when not under fire."
```
Comment (Dorsisdwarf) [inline code/modules/clothing/spacesuits/hardsuit.dm]: 20s per charge isn't exactly rapidly.

File: code/modules/clothing/spacesuits/hardsuit.dm [hunk 4]
```diff
@@ -430,3 +430,76 @@
...
+	item_state = "hardsuit0-ert_security"
+	item_color = "ert_security"
+	helmettype = /obj/item/clothing/head/helmet/space/hardsuit/ert/sec
+	armor = list(melee = 0, bullet = 30, laser = 30, energy = 30, bomb = 50, bio = 100, rad = 100)
+	slowdown = 0
+
+/obj/item/clothing/suit/space/hardsuit/shielded/ctf/blue
+	name = "blue shielded hardsuit"
+	desc = "Standard issue hardsuit for playing capture the flag."
+	icon_state = "ert_command"
```
Comment (Dorsisdwarf) [inline code/modules/clothing/spacesuits/hardsuit.dm]: >missing the opportunity for references

Comment (KorPhaeron) [inline code/modules/clothing/spacesuits/hardsuit.dm]: Gimme some good ones then

File: code/modules/clothing/spacesuits/hardsuit.dm (hunk 2 above)
Comment (duncathan) [inline code/modules/clothing/spacesuits/hardsuit.dm]: In fact the style guidelines in the code standards mentions this...
Dammit kor

File: code/modules/clothing/spacesuits/hardsuit.dm (hunk 4 above)
Comment (duncathan) [inline code/modules/clothing/spacesuits/hardsuit.dm]: I love halo

Comment (Dorsisdwarf) [inline code/modules/clothing/spacesuits/hardsuit.dm]: Standard issue hardsuit for having a base in a box canyon.

**Comment by Steelpoint** [comment]
Interesting attempt at making a real sort of advance power armour, however not all damage is made equally.

What constitutes a 'hit' for the suit to defend against? Does it have to do damage or can you throw a no-damaging item at someone to lower their shield? Overall a minor issue since either way someone will either waste time throwing pens at you or expend more precious ammo at you.

Also I'll be interested in how you intend for this to be used on station if at all.

**Comment by Niknakflak** [comment]
This reminds me of that armor in legend of zelda twillight princess where you were invincible until you ran out of rubys

**Comment by KorPhaeron** [comment]
If only we had economy

**Comment by KorPhaeron** [comment]
@Steelpoint

Treats every disarm/gunshot/melee blow the same. I left the code easy to update to take damage into account, but tasers do 0 damage and probably shouldnt be ignored.

Only vague plan I have for on station is maybe putting a (one-two hit) shield in the captain or hos spacesuits.

This will probably just be thunderdome stuff for now though

**Comment by KorPhaeron** [comment]
Thanks to @RemieRichards this is done now, even if only as thunderdome/admin items/proof of concept.

**Comment by Iamgoofball** [comment]
ima put it in R&D

**Comment by KorPhaeron** [comment]
Weird, it compiles locally.

**Comment by KorPhaeron** [comment]
@Iamgoofball 

Surely we can think of a better use for these than just dumping them into the loot pile that is R&D

**Comment by Iamgoofball** [comment]
@KorPhaeron 

cargo it is

**Comment by Zilenan91** [comment]
Yeah don't let science get even more of the cool shit.

Alternatively, we can make ghetto assistant power armor out of hazard vests and power cells that have a 50% chance to either work as intended or shock the wearer for damage.

**Comment by MMMiracles** [comment]
Maybe replace the ablative armor with a variant of this so it actually makes sense to be called ablative armor?

**Comment by Zilenan91** [comment]
I wouldn't mind if we added the shield on top of the protection it already has.

**Comment by duncathan** [comment]
@KorPhaeron did you forget to commit effects.dmi? That's what Travis is angry about

**Comment by KorPhaeron** [comment]
@duncathan 

I ended up not editing effects at all, guess I will try though. Might just remake the PR

**Pull request closed without merging**